lock/etcd.go

```diff
@@ -60,46 +64,54 @@ func NewEtcdLockClient(keyapi KeysAPI, group string) (*EtcdLockClient, error) {
 
 	elc := &EtcdLockClient{keyapi, key}
 	if err := elc.Init(); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("unable to init etcd lock client: %w", err)
 	}
 
 	return elc, nil
 }
 
 // Init sets an initial copy of the semaphore if it doesn't exist yet.
+// So we first try to get the value, if the value is not found we create the key
+// with a default semaphore value
 func (c *EtcdLockClient) Init() error {
-	sem := newSemaphore()
-	b, err := json.Marshal(sem)
-	if err != nil {
-		return err
-	}
+	if _, err := c.Get(); err != nil {
+		if errors.Is(err, ErrNotFound) {
+			sem := newSemaphore()
+			if err := c.Set(sem); err != nil {
+				return fmt.Errorf("unable to set semaphore: %w", err)
+			}
```

Comment:
Yes that would be totally possible - in case you have a couple of Flatcar instances consuming the same `etcd` cluster, we could have a race condition like the one you mentioned.
From what I understood, the first `c.Set` in the `Init` method will create an empty `Semaphore` - if a second client tries to create another empty `Semaphore` it will just increase the `Version` attribute of the previous semaphore; so in the end we have the same behavior, we have created one `empty` semaphore to `init` the `etcd` KV store and since in the `c.Get` we always fetch the latest created value: the previously created values won't interfere.

Reply:
So I have a follow-up question: What would happen in the following scenario:
![locksmith-case](https://user-images.githubusercontent.com/96081/129731537-919d8cec-b0a8-4dcd-af4d-e6d06fff2cda.png)

I'm wondering if using [`Txn`](https://pkg.go.dev/go.etcd.io/etcd/client/v3#Txn) would be the way to go. [Docs about migrating from v2 to v3](https://etcd.io/docs/v3.5/op-guide/v2-migration/) seem to be rather poor…

Reply:
y the doc is pretty poor it only mentions the migration from the app side and not from the lib side. The `txn` seems to correctly handle this case.

What do you think about this transaction - if the version of `k1` is > 1 (which means the initial semaphore is already there) we just do nothing and on the other hand, we put our empty semaphore:

```go
Txn(context.TODO()).If(
 Compare(Version(k1), ">", 1)
).Then(
 // do nothing
).Else(
 OpPut(k5,semaphore)
).Commit()
```

Reply:
> y the doc is pretty poor it only mentions the migration from the app side and not from the lib side. The `txn` seems to correctly handle this case.
> 
> What do you think about this transaction - if the version of `k1` is > 1 (which means the initial semaphore is already there) we just do nothing and on the other hand, we put our empty semaphore:

Why not greater than 0, instead of 1? Wouldn't a version greater than 0 already mean that there is a semaphore? [There's a comment that would suggest it](https://github.com/etcd-io/etcd/issues/6518#issuecomment-249622279).

> 
> ```go
> Txn(context.TODO()).If(
>  Compare(Version(k1), ">", 1)
> ).Then(
>  // do nothing
> ).Else(
>  OpPut(k1,semaphore)
> ).Commit()
> ```

If we would go with greater than zero then maybe this could become a bit simpler when we reverse the condition to move "else" branch to "then", and then dropping the now-empty "else" branch.

```go
Txn(context.TODO()).If(
  Compare(Version(k1), "=", 0)
).Then(
  OpPut(k1, semaphore)
).Commit()
```

Reply:
@krnowak I pushed - updated the tests. Out of curiosity, which tool did you use to generate this graph ? :arrow_up:

Reply:
It's an indeed [nifty tool](https://excalidraw.com/). :)